validator/db/proposal_history.go

```diff
@@ -0,0 +1,101 @@
+package db
+
+import (
+	"github.com/boltdb/bolt"
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+	slashpb "github.com/prysmaticlabs/prysm/proto/slashing"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+// HasProposedForEpoch returns whether a validators proposal history has been marked for the entered epoch.
+// If the request is more in the future than what the history contains, it will return false.
+// If the request is from the past, and likely previously pruned it will return true to avoid slashing.
+func HasProposedForEpoch(history *slashpb.ProposalHistory, epoch uint64) bool {
+	wsPeriod := params.BeaconConfig().WeakSubjectivityPeriod
+	// Previously pruned, but to be safe we should return true.
+	if int(epoch) <= int(history.LatestEpochWritten)-int(wsPeriod) {
+		return false
+	}
+	// Accessing future proposals that haven't been marked yet. Needs to return false.
+	if epoch > history.LatestEpochWritten {
+		return false
+	}
+	return history.EpochBits.BitAt(epoch % wsPeriod)
+}
+
+// SetProposedForEpoch updates the proposal history to mark the indicated epoch in the bitlist
+// and updates the last epoch written if needed.
+func SetProposedForEpoch(history *slashpb.ProposalHistory, epoch uint64) {
```

Comment:
Should these functions be here with the DB?

Reply:
yes, this function describes a write to your proposal history, so you do want to save it to the db. At the end of this method you should then call `SaveProposalHistory`